File: php/src/Resource/Workflows.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\StartRun;
use Fivexer\SDK\Model\WorkflowDefinition;
use Fivexer\SDK\Model\WorkflowDefinitionInput;
use Fivexer\SDK\Model\WorkflowDefinitionSummary;
use Fivexer\SDK\Model\WorkflowRun;
use Fivexer\SDK\Model\WorkflowRunPage;

/** Workflow definitions, and starting runs from them. */
final class Workflows
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    /** @return list<WorkflowDefinitionSummary> */
    public function list(): array
    {
        $data = $this->client->request('GET', '/workflows') ?? [];
        /** @var list<WorkflowDefinitionSummary> */
        return Json::parseEach($data, 'workflows', [WorkflowDefinitionSummary::class, 'fromArray']);
    }

    public function get(string $workflowId): WorkflowDefinition
    {
        return WorkflowDefinition::fromArray(
            $this->client->request('GET', '/workflows/' . \rawurlencode($workflowId)) ?? []
        );
    }

    /**
     * Create or replace a definition. The engine validates the graph on save (unreachable or
     * dangling steps, a missing initial step, a machine step without a handler), so a bad
     * definition comes back as 400 invalid_workflow.
     */
    public function save(string $workflowId, WorkflowDefinitionInput $input): WorkflowDefinition
    {
        return WorkflowDefinition::fromArray(
            $this->client->request('PUT', '/workflows/' . \rawurlencode($workflowId), $input->toArray()) ?? []
        );
    }

    public function remove(string $workflowId): void
    {
        $this->client->request('DELETE', '/workflows/' . \rawurlencode($workflowId));
    }

    public function run(string $workflowId, ?StartRun $input = null): WorkflowRun
    {
        $body = ($input ?? new StartRun())->toArray();
        return WorkflowRun::fromArray(
            $this->client->request('POST', '/workflows/' . \rawurlencode($workflowId) . '/runs', $body) ?? []
        );
    }

    /**
     * Runs of one definition, newest-first and cursor-paginated. The workflow id is already in
     * the path, so it is deliberately not repeated as a filter.
     */
    public function listRuns(
        string $workflowId,
        ?string $status = null,
        ?string $cursor = null,
        ?int $limit = null,
    ): WorkflowRunPage {
        $data = $this->client->request('GET', '/workflows/' . \rawurlencode($workflowId) . '/runs', null, [
            'status' => $status,
            'cursor' => $cursor,
            'limit' => $limit,
        ]) ?? [];
        return WorkflowRunPage::fromArray($data);
    }
}

File: java/src/main/java/io/fivexer/sdk/Workflows.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.StartRun;
import io.fivexer.sdk.model.WorkflowDefinition;
import io.fivexer.sdk.model.WorkflowDefinitionInput;
import io.fivexer.sdk.model.WorkflowDefinitionSummary;
import io.fivexer.sdk.model.WorkflowList;
import io.fivexer.sdk.model.WorkflowRun;
import io.fivexer.sdk.model.WorkflowRunPage;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/** Workflow definitions, and starting runs from them. */
public final class Workflows {

    private final Fivexer client;

    Workflows(Fivexer client) {
        this.client = client;
    }

    public List<WorkflowDefinitionSummary> list() {
        WorkflowList envelope = client.request("GET", "/workflows", null, null, WorkflowList.class, false);
        return envelope.getWorkflows() == null ? Collections.emptyList() : envelope.getWorkflows();
    }

    public WorkflowDefinition get(String workflowId) {
        return client.request("GET", "/workflows/" + Json.enc(workflowId), null, null,
                WorkflowDefinition.class, false);
    }

    /**
     * Create or replace a definition. The engine validates the graph on save (unreachable or
     * dangling steps, a missing initial step, a machine step without a handler), so a bad
     * definition comes back as {@code 400 invalid_workflow}.
     */
    public WorkflowDefinition save(String workflowId, WorkflowDefinitionInput input) {
        return client.request("PUT", "/workflows/" + Json.enc(workflowId), input.toJson(), null,
                WorkflowDefinition.class, false);
    }

    public void remove(String workflowId) {
        client.request("DELETE", "/workflows/" + Json.enc(workflowId), null, null, Void.class, true);
    }

    public WorkflowRun run(String workflowId, StartRun input) {
        StartRun start = input == null ? new StartRun() : input;
        return client.request("POST", "/workflows/" + Json.enc(workflowId) + "/runs", start.toJson(),
                null, WorkflowRun.class, false);
    }

    public WorkflowRun run(String workflowId) {
        return run(workflowId, null);
    }

    /** Runs of one definition, newest-first and cursor-paginated. */
    public WorkflowRunPage listRuns(String workflowId, ListRunsQuery query) {
        ListRunsQuery q = query == null ? new ListRunsQuery() : query;
        // The workflow id is already in the path; repeating it as a filter would be noise.
        Map<String, String> filters = q.toQuery();
        if (filters != null) {
            filters.remove("workflowId");
            if (filters.isEmpty()) {
                filters = null;
            }
        }
        return client.request("GET", "/workflows/" + Json.enc(workflowId) + "/runs", null, filters,
                WorkflowRunPage.class, false);
    }

    public WorkflowRunPage listRuns(String workflowId) {
        return listRuns(workflowId, null);
    }
}
